File: core/amber/src/main/python/core/architecture/handlers/store_upstream_link_ids_handler.py
from proto.edu.uci.ics.amber.engine.architecture.worker import StoreUpstreamLinkIdsV2
from .handler_base import Handler
from ..managers.context import Context


class StoreUpstreamLinkIdsHandler(Handler):
    cmd = StoreUpstreamLinkIdsV2

    def __call__(self, context: Context, command: StoreUpstreamLinkIdsV2, *args, **kwargs):
        context.batch_to_tuple_converter.update_all_upstream_link_ids(
            set(command.upstream_link_ids)
        )
        return None

File: core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/promisehandlers/StoreUpstreamLinkIdsHandler.scala
package edu.uci.ics.amber.engine.architecture.worker.promisehandlers

import edu.uci.ics.amber.engine.architecture.worker.WorkerAsyncRPCHandlerInitializer
import edu.uci.ics.amber.engine.architecture.worker.promisehandlers.StoreUpstreamLinkIdsHandler.StoreUpstreamLinkIds
import edu.uci.ics.amber.engine.common.rpc.AsyncRPCServer.ControlCommand
import edu.uci.ics.amber.engine.common.virtualidentity.LinkIdentity

object StoreUpstreamLinkIdsHandler {

  final case class StoreUpstreamLinkIds(upstreamLinkIds: Array[LinkIdentity])
      extends ControlCommand[Unit]
}

/**
  * Used to tell the worker how many upstream links it will get the data from
  * Sender: Controller
  */
trait StoreUpstreamLinkIdsHandler {
  this: WorkerAsyncRPCHandlerInitializer =>

  registerHandler { (msg: StoreUpstreamLinkIds, sender) =>
    batchToTupleConverter.updateAllUpstreamLinkIds(msg.upstreamLinkIds.toSet)
  }

}
